step3 리뷰 요청드립니다.

deokmoon opened the pull request:
안녕하세요 영훈님 :)

오랜만에 인사드리는 것 같습니다.
뭔가 늘 정신없이 미션을 진행 하는 것 같아 올바른 방향으로 하고 있는지 잘 모르겠습니다. 😢 
(개인적인 사정으로 좀 늦어져서 조급한 마음도 있어 더 그런 것 같습니다.)

이에 미션을 섞어 진행하면서, 어디까지 했지? 라는 생각이 들면서 진행하니 실수도 많은 것 같더라고요.
그래서 최대한 readme에 todoList를 상세히 적어, todo 리스트를 해결하는 방향으로 진행하려고 했습니다. 

이전 atdd 미션에서는 영훈님이 리뷰 주신 것처럼 body에 모델링한 request 객체를 담아 호출하는 것으로 수정하였습니다.
예제에서 map으로 활용하기에, 당연히 map으로 처리해야하는 것은 아닌가? 라는 생각을 가졌었습니다.
설계한 request 객체를 활용하니 map에 put 하는 과정이 필요 없어져 더 깔끔해졌네요
리뷰 감사드립니다!

이번 step3에서는 Line -  Sections - Section - Station 을 관계로 미션을 수행해보았습니다. 
아직 JPA가 서툴러 지금한 연관관계 매핑이 미션에서 요구한 내용인지 확신은 들지 않지만, 
미션에서 요구한  다대다 매핑을 피하기 위해 Section 엔티티를 두었습니다.
이러한 Section을 List로 갖는 Sections 일급콜렉션을 생성하여 Sections의 기능을 처리를 위한 책임을 일임하였습니다.

Step3도 정신없이 진행해서 모자란 점이 많겠지만, 리뷰 잘 부탁드리겠습니다!


ps. 혹시 restassured를 검색하면 아직 사람들이 정리해놓은 자료가 별로 없는데요. 이럴 떄는 주로 오피셜 docs 로 확인하시나요?
(https://rest-assured.io/)

deokmoon committed:
refactor: step2 feedback 적용 - @transactional 내의 modify 호출 불필요

deokmoon committed:
refactor: step2 feedback 적용 - 불필요한 일급콜렉션 제거에 따린 리팩토링

deokmoon committed:
refactor: step2 feedback 적용 - RestAssured request 객체로 리팩토링

deokmoon committed:
refactor: step2 feedback 적용 - TestFactory 적용

deokmoon committed:
docs: step2 회고 및 step3 요구사항 작성

deokmoon committed:
docs: 기능 요구사항 구체적으로 작성

deokmoon committed:
docs: 기능 요구사항 수정

deokmoon committed:
test: Sections 단위테스트 작성

deokmoon committed:
feat: Sections add 예외처리 및 upStation 동일한 경우 구현

deokmoon committed:
feat: add section api 처리를 위한 각 프로덕트 코드 작성

deokmoon committed:
test: 인수테스트 작성 및 공통 부분 부모클래스로 변경

deokmoon committed:
feat: 조회 기능 추가 구현

deokmoon committed:
feat: section, sections 프로덕트 코드 작성

Commit:
refactor: 요구사항 내용 참고하여 인수기반 테스트 수정

deokmoon committed:
refactor: 미사용으로 삭제 처리

deokmoon committed:
feat: 모두 일치하지 않은 지하철역인 경우 예외처리 적용

deokmoon committed:
docs: 요구사항 재정리

deokmoon committed:
refactor: factory static method 적용으로 인한 리팩토링

deokmoon committed:
refactor: 불필요한 repository 삭제

deokmoon committed:
refactor: sectionService로 분리

deokmoon committed:
test: 상행,하행 일치 및 예외사항에 대한 테스트코드 작성

deokmoon committed:
feat: 종점 관련 기능 1차 구현

deokmoon committed:
test: 상하행 정렬 테스트 코드 작성

deokmoon committed:
feat: 상하행 정렬관련 프로덕트 코드 작성

deokmoon committed:
test: 종점 추가 인수 테스트코드 작성

chae-yh commented on README.md:
```diff
@@ -151,22 +151,72 @@ Vary: Access-Control-Request-Headers
   * Given 지하철 노선을 생성하고
   * When 생성한 지하철 노선을 삭제하면
   * Then 해당 지하철 노선 정보는 삭제된다
+#### Step2 회고
```
기능 요구 사항 작성 👍 저번에도 말씀드리긴 했지만, 회고를 작성하는 부분도 좋네요 ㅎㅎ 👍 

추가로, 질문 주신 내용도 여기서 답변 드리겠습니다.

`뭔가 늘 정신없이 미션을 진행 하는 것 같아 올바른 방향으로 하고 있는지 잘 모르겠습니다. 😢`

제 생각에는 잘 진행해주시는 것 같습니다. 과정 자체가 많은 작업량을 요구하고, 업무도 병행하고 계시는 만큼 다른 분들도 시간없이, 혹은 정신없이 진행해주시는 경우가 많은 것 같습니다. ㅎㅎㅎ

`혹시 restassured를 검색하면 아직 사람들이 정리해놓은 자료가 별로 없는데요. 이럴 떄는 주로 오피셜 docs 로 확인하시나요?`

넵 말씀하신대로, 해당 기술을 만든 곳에서 제공하는 메뉴얼이 가장 정확하고 가장 내용이 풍부하다고 생각되어 해당 자료를 참조하고, 실제 동작을 확인할 때 코드 레벨로 보거나, 혹은 코드 내에도 설명이 잘되어 있는경우가 많아 코드 자체를 참조하기도 합니다. 
![image](https://user-images.githubusercontent.com/81790362/205415391-f61222a2-c388-44f7-aec9-55af81470a10.png)

그리고 여러가지 유스케이스에 대해서는 [Stackoverflow](https://stackoverflow.com/)도 많이 참조합니다.

deokmoon replied:
상세한 답변 감사드립니다!

chae-yh commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -21,45 +22,53 @@ public LineService(LineRepository lineRepository, StationRepository stationRepos
         return LineResponse.from(lineRepository.save(lineRequest.toLine(upStation, downStation)));
     }
 
     public List<LineResponse> findAllLines() {
```
이렇게 조회만하는 경우에는 @Transactional(readOnly = true) 옵션을 주는 것도 좋을 것 같아요..!

chae-yh replied:
제가 못찾는 것인지.. 적용이 안되어 있는 것 같아요...!

deokmoon replied:
엇.. 해당 메서드가 아닌, 클래스에 공통적으로  @transactional(readOnly = true) 적용 후,
teadOnly true가 아닌 저장(수정), delete, 과 같은 경우에만 메서드 @transactional을 적용하였습니다..

chae-yh commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -40,4 +53,114 @@ public void belongLine(Line line) {
+    }
+
+    public boolean isSameUpStationId(Section section) {
+        return upStation.getId() == section.findUpStationId();
```
Entity이므로, 해당 형태처럼 id만 비교해도 될 것 같습니다..! 다만, 이런 비교는 Station의 Equals 와 HashCode 를 재정의해서 비교해보면 어떨까요...?

deokmoon replied:
나중에 equals hashcode 재정의 해놓고 이 부분에서는 == 비교했네요

chae-yh commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -40,4 +53,114 @@ public void belongLine(Line line) {
+    }
+
+    private void checkDistanceValidation(Long existingDistance, Long newDistance) {
+        if (existingDistance < newDistance || existingDistance == newDistance) {
```
```suggestion
        if (existingDistance <= newDistance) {
```
이렇게 작성해줄 수 있지 않을까요..?

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -25,10 +28,121 @@ public Sections(List<Section> sections) {
+
+    private boolean isSameUpStation(Section section) {
+        return sections.stream()
+                .filter(it -> it.isSameUpStationId(section))
```
it 같은 축약어보다는 변수를 잘 설명해줄 수 있는 이름을 주는 것은 어떨까요...?

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -25,10 +28,121 @@ public Sections(List<Section> sections) {
+        if (sections.stream().anyMatch(s -> s.equals(section))) {
+            throw new IllegalArgumentException(ErrorCode.NO_SAME_SECTION_EXCEPTION.getErrorMessage());
+        }
+        if (this.sections.size() > 0 && isNoMatchStation(section)) {
```
0은 상수 추출해도 좋을 것 같아요..!

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -25,10 +28,121 @@ public Sections(List<Section> sections) {
+                .findFirst().isPresent();
+    }
+
+    private void checkValidation(Section section) {
```
기존에 구간이 A-B-C-D-E가 있었는데, B-D라는 구간이 들어와도 이 validation에서 검증이 될까요...? (해당 구간은 만들어지면 안됨)

deokmoon replied:
하나의 섹션만 비교했고, 기존에 이미 있는 station인지에 대한 검증이 없었네요!
놓친 부분이네요 리뷰 감사합니다!

chae-yh commented on src/test/java/nextstep/subway/AcceptanceTest.java:
```diff
@@ -14,4 +22,47 @@ public class AcceptanceTest {
     protected DatabaseCleaner databaseCleaner;
+
+    @Autowired
+    StationService stationService;
```
해당 클래스는 인수테스트의 중복된 부분을 추출한 클래스로 보이는데, StationService는 도메인에 가까운 클래스로 보입니다...! 이 부분은 StationAccpetanceFixture 등으로 분리해보면 어떨까요..?
추가로, 접근 제한자는 가능한 작성하는 편이 좋을 것 같습니다..! 작성하지 않으면 protected이므로 명시해주는 것이 코드 읽을 떄 좋을 것 같아요 ㅎㅎ

chae-yh commented on src/test/java/nextstep/subway/Line/LineAcceptanceTest.java:
```diff
@@ -45,15 +39,20 @@ public void setUp() {
-        // then
-        assertThat(lineNames).containsAnyOf("2호선");
+    @TestFactory
+    Stream<DynamicTest> createLine() {
```
Dynamic Test 👍

chae-yh commented on src/test/java/nextstep/subway/Section/SectionsAcceptanceTest.java:
```diff
@@ -0,0 +1,138 @@
+import static org.junit.jupiter.api.DynamicTest.dynamicTest;
+
+@DisplayName("지하철 노선 관련 기능")
+public class SectionsAcceptanceTest extends AcceptanceTest {
```
Acceptance Test 꼼꼼히 잘 작성된 것 같습니다 👍

chae-yh commented on src/test/java/nextstep/subway/Section/SectionsTest.java:
```diff
@@ -0,0 +1,93 @@
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+@DataJpaTest
+class SectionsTest {
```
Sections에 대한 단위 테스트 👍 다만 `@DataJpaTest는 왜 들어갔는지 여쭤봐도 될까요...?

deokmoon replied:
의도는 쿼리 잘 나가나 로그 찍히는거 보려고 했었습니다.

chae-yh replied:
데이터베이스에 대한 테스트가 아니여서 `@DataJpaTest는 필요가 없을 것 같습니다. 삭제하는 것은 어떨까요...?

deokmoon replied:
아! 그런 의미로 테스트코드를 작성하는게 좋겠네요 👍 
리뷰 감사합니다

chae-yh reviewed (changes requested):
덕문님 안녕하세요. 피드백이 다소 늦어지게 되어 죄송합니다. 
그리고 코드 작성은 전체적으로 잘 해주셨습니다 👍 문서 정리를 잘해주셔서 읽기가 편하네요 ㅎㅎ 다만 고민이 조금 필요한 부분이 있어 코멘트 드리니 확인 부탁드립니다.

감사합니다 : )

deokmoon committed:
refactor: 1차 리뷰 적용
 - filter 변수명 it -> eachSection으로 적용
 - transactional readonly 표시
 - 공통 인수테스트에서 testfixture 분리

deokmoon committed:
fix: readOnly 클래스 적용 및 메서드에 transactional 적용

deokmoon committed:
teat: station 중복 테스트코드 작성

deokmoon committed:
feat: station 중복 체크하는 validation 추가 적용

chae-yh commented on src/test/java/nextstep/subway/AcceptanceTest.java:
```diff
@@ -14,4 +22,24 @@ public class AcceptanceTest {
+        databaseCleaner.execute();
+    }
+
+    public StationRequest generateStationRequest(String name) {
```
이 메서드와 아래의 메서드들도 전부 도메인 테스트로 가는 편이 좋지 않을까요...?

deokmoon replied:
음.. 2군데 이상의 도메인 테스트에서 사용해서 공통용으로 넣었는데,
다시 생각해보니 도메인 영역에 있는 것이 적절해 보이네요!

chae-yh commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -40,4 +53,114 @@ public void belongLine(Line line) {
+        return upStation.equals(section.getUpStation());
+    }
+
+    public boolean isSameDownStationId(Section section) {
```
id로 비교하고 있지 않아서, 메서드 명을 변경해주어도 좋을 것 같습니다..!

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -25,10 +30,152 @@ public Sections(List<Section> sections) {
+        boolean hasIssue = false;
+        Section firstSection = findFirstSection();
+        Section lastSection = findLastSection();
+        if (firstSection.isEqualUpStationNewSectionDownStation(newSection)
```
```suggestion
        return firstSection.isEqualUpStationNewSectionDownStation(newSection) && lastSection.isEqualDownStationNewSectionUpStation(newSection)
```

hasIssue 없이 이렇게만 해주어도 되지 않을까요...?

deokmoon replied:
오 넵  boolean 지역 변수도 제거되고 좋네요!  리뷰 감사드립니다.

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -25,10 +30,152 @@ public Sections(List<Section> sections) {
+
+
+    private void updateStationWhenDownStationSame(Section section) {
+        Optional<Section> findSection = sections.stream().filter(eachSection -> eachSection.isSameDownStationId(section)).findFirst();
```
ifPresent로 한번에 써줄 수도 있을 것 같아요..!

chae-yh commented on src/test/java/nextstep/subway/station/StationAcceptanceFixture.java:
```diff
@@ -0,0 +1,17 @@
+    protected StationService stationService;
+
+    @Transactional
+    public long creatStation(String name) {
```
StationService의 saveStation을 사용하면 될 것 같은데, 혹시 따로 만들어주신 이유가 있을까요...?

deokmoon replied:
아 그러는게 깔끔하겠네요 ! 아직 JPA가 친숙치 않아 시야가 좁은 것 같습니다.
리뷰 감사합니다!

chae-yh reviewed (approved):
덕문님 안녕하세요. 피드백 드린 부분 잘 반영해주신 것 같습니다 👍 조금 더 고민해주시면 좋을만한 내용들이 있지만, 전체적으로 미션의 목적은 달성했다고 생각하여 merge 하겠습니다. 피드백 드린 내용은 다음 step 진행하시면서 살짝 확인해주시면 될 것 같아요..!

이제 마지막 step만 남았네요. 파이팅 입니다 : )
